src/safeposix/syscalls/fs_calls.rs

```diff
@@ -1537,4 +1536,105 @@ impl Cage {
+                            // if length is greater than original filesize,
+                            // file is extented with null bytes
+                            if filesize < length {
+                                let blankbytecount: usize = length - filesize;
+                                
+                                // create a null-byte buffer of proper size
+                                let mut vec = vec![0u8; blankbytecount];
+                                let buf: *mut u8 = &mut vec[0];
+
+                                // extend file with null-byte buffer
+                                if let Ok(byteswritten) = fileobject.writeat(buf, blankbytecount, filesize) {
+                                    // update filesize if writeat() succeeded
+                                    normalfile_inode_obj.size = length;
+                                } else {
+                                    panic!("writeat() failed");
+                                }
+                            } else { // if length is smaller than original filesize,
+                                     // extra data are cut off
+                                // create a null-byte buffer of proper size
+                                let mut vec = vec![0u8; length];
+                                let buf: *mut u8 = &mut vec[0];
+                                
+                                // read the part to keep to properly sized buffer
+                                if let Ok(bytesread) = fileobject.readat(buf, length, 0) {
+                                    // remove file from file object table if readat() succeeded
+                                    fobjtable.remove(&inodenum).unwrap().close().unwrap();
+                                    let filename = format!("{}{}", FILEDATAPREFIX, inodenum);
+                                    interface::removefile(filename).unwrap();
+
+                                    // insert new file created using old filename to file object table
+                                    let filename1 = format!("{}{}", FILEDATAPREFIX, inodenum);
+                                    fobjtable.insert(inodenum, interface::openfile(filename1, true).unwrap());
+                                    fileobject = fobjtable.get_mut(&inodenum).unwrap();
+
+                                    // write the kept part to new file from buffer
+                                    if let Ok(byteswritten) = fileobject.writeat(buf, length, 0) {
+                                        // update filesize if writeat() succeeded
+                                        normalfile_inode_obj.size = length;
+                                    } else {
+                                        panic!("writeat() failed");
+                                    }
+                                } else {
+                                   panic!("readat() failed");
+                                }
```

Comment:
The fact that we did it like this in repy safeposix is because RePy had no way to truncate the system file as it did not access the truncate syscall. We may not want to keep it this way because it seems ridiculously expensive and not very smart to do it this way. @rennergade what are your thoughts?

Reply:
I've added the function `shrink` to the interface, which is a method for EmulatedFile, this should easily replace all this code so just merge it in.